source/h2_connection.c

```diff
@@ -735,16 +772,46 @@ static int s_handler_process_read_message(
     struct aws_channel_handler *handler,
     struct aws_channel_slot *slot,
     struct aws_io_message *message) {
-
-    (void)handler;
     (void)slot;
-    (void)message;
+    struct aws_h2_connection *connection = handler->impl;
+
+    CONNECTION_LOGF(
+        TRACE,
+        connection,
+        "id=%p: H2 connection Begin processing message of size %zu.",
+        (void *)&connection->base,
+        message->message_data.len);
+
+    if (connection->thread_data.is_reading_stopped) {
+        CONNECTION_LOGF(
+            ERROR,
+            connection,
+            "id=%p: Cannot process message because connection is shutting down.",
+            (void *)&connection->base);
 
+        aws_raise_error(AWS_ERROR_HTTP_CONNECTION_CLOSED);
+        goto shutdown;
+    }
     /* HTTP/2 protocol uses WINDOW_UPDATE frames to coordinate data rates with peer,
      * so we can just keep the aws_channel's read-window wide open */
+    struct aws_byte_cursor message_cursor = aws_byte_cursor_from_buf(&message->message_data);
+
     /* #TODO update read window by however much we just read */
+    aws_h2_decode(connection->thread_data.decoder, &message_cursor);
```

Comment:
check aws_h2_decode() for errors

also, move the decoding stuff up above the comments about "read-window"